source/common/network/connection_impl.cc

```diff
@@ -395,6 +396,21 @@ void ConnectionImpl::updateWriteBufferStats(uint64_t num_written, uint64_t new_s
                                            buffer_stats_->write_current_);
 }
 
+const std::string Network::ConnectionImpl::destinationAddress() {
+  if (fd_ != -1) {
+    sockaddr_storage orig_dst_addr;
+    memset(&orig_dst_addr, 0, sizeof(orig_dst_addr));
+    bool success = Utility::getOriginalDst(fd_, &orig_dst_addr);
+    if (success) {
+      return Utility::urlForTcp(
+          Utility::getAddressName(reinterpret_cast<struct sockaddr_in*>(&orig_dst_addr)),
+          Utility::getAddressPort(reinterpret_cast<struct sockaddr_in*>(&orig_dst_addr)));
+    }
+  }
+
+  return EMPTY_STRING;
```

Comment:
The reason why I didn't do that was to avoid paying the cost of the getsockopt() system call if it is not needed. This would be the majority of the cases, since we only need that if we use L4 routing based on destination IP/port. If you think it might be useful for other purposes, we can do it in the constructor like you said and also have localAddress() return a const &.